### PR TITLE
Added support to show loading status for the last step of OAuth code flow

### PR DESCRIPTION
**Pull request by xuyanke**
OAuth code flow requires a network request to get access token, apps should build some UX to handle this async operation.
A default UX is built in the SDK with showing a load spinner while waiting for oauth token request to finish.
Apps can provide a `LoadingStatusDelegate` to override the default and show a custom experience.

### CODE REVIEW

File: Source/SwiftyDropbox/Platform/SwiftyDropbox_iOS/OAuthMobile.swift [hunk 1]
```diff
@@ -32,13 +41,17 @@ extension DropboxClientsManager {
...
     ///     - sharedApplication: The shared UIApplication instance in your app.
     ///     - controller: A UIViewController to present the auth flow from.
+    ///     - loadingStatusDelegate: An optional delegate to handle loading experience during auth flow.
+    ///       e.g. Show a looading spinner and block user interaction while loading/waiting.
+    ///       If a delegate is not provided, the SDK will show a default loading spinner when necessary.
     ///     - openURL: Handler to open a URL.
     ///     - scopeRequest: Contains requested scopes to obtain.
     public static func authorizeFromControllerV2(
-        _ sharedApplication: UIApplication, controller: UIViewController?, openURL: @escaping ((URL) -> Void), scopeRequest: ScopeRequest?
+        _ sharedApplication: UIApplication, controller: UIViewController?, loadingStatusDelegate: LoadingStatusDelegate?, openURL: @escaping ((URL) -> Void), scopeRequest: ScopeRequest?
```
Comment (rafe-g) [inline Source/SwiftyDropbox/Platform/SwiftyDropbox_iOS/OAuthMobile.swift:50]: is this delegate ever non nil? i see the usage here passes in a nil instance https://github.com/dropbox/SwiftyDropbox/pull/271/files#diff-47b40edc806c91305e9791514ebf7eceR24

File: Source/SwiftyDropbox/Platform/SwiftyDropbox_iOS/OAuthMobile.swift [hunk 2]
```diff
@@ -320,6 +335,33 @@ open class MobileSharedApplication: SharedApplication {
...
+            let loadingView = safariViewController.presentedViewController as? LoadingViewController {
+            // Web OAuth flow, dismiss loading view on the MobileSafariViewController.
+            loadingView.dismiss(animated: false)
+        } else {
+            if let loadingStatusDelegate = loadingStatusDelegate {
+                loadingStatusDelegate.dismissLoading()
+            } else if let loadingView = controller?.presentedViewController as? LoadingViewController {
+                loadingView.dismiss(animated: false)
+            }
+        }
```
Comment (rafe-g) [inline Source/SwiftyDropbox/Platform/SwiftyDropbox_iOS/OAuthMobile.swift:363]: This logic is pretty hard to follow. Could we have a helper computed property that returns an optional loading controller ? That would would simplify logic like:

```
var loadingController: LoadingViewController? {
    let safariController = controller?.presentedViewController as? MobileSafariViewController
    return safariViewController.presentedViewController as? LoadingViewController ?? controller?.presentedViewController as? LoadingViewController
}
```

This method would then simplify to:


```
    if let loadingController = self.loadingController {
        loadingController.dismiss(animated: false)
    } else {
        loadingStatusDelegate?.dismissLoading()
    }
```

This is nicer as it's exhaustive and you dont have all these weird casting going on.

File: Source/SwiftyDropbox/Shared/Handwritten/OAuth/OAuth.swift [hunk 3]
```diff
@@ -31,6 +33,7 @@ open class DropboxOAuthManager {
     /// Session data for OAuth2 code flow with PKCE.
     /// nil if we are in the legacy token flow.
     var authSession: AuthSession?
+    weak var sharedApplication: SharedApplication?
```
Comment (rafe-g) [inline Source/SwiftyDropbox/Shared/Handwritten/OAuth/OAuth.swift:36]: why is this weak?

File: Source/SwiftyDropbox/Platform/SwiftyDropbox_iOS/LoadingViewController.swift [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+import UIKit
+
+/// A VC with a loading spinner at its view center.
+class LoadingViewController: UIViewController {
+    private let loadingSpinner: UIActivityIndicatorView
+
+    override init(nibName nibNameOrNil: String?, bundle nibBundleOrNil: Bundle?) {
+        loadingSpinner = UIActivityIndicatorView(style: .whiteLarge)
+        super.init(nibName: nibNameOrNil, bundle: nibBundleOrNil)
+        modalPresentationStyle = .overFullScreen
```
[the remark this answers is not in the document]
Comment (xuyanke) [reply Source/SwiftyDropbox/Platform/SwiftyDropbox_iOS/LoadingViewController.swift:15]: This has to be full screen to block user interaction. You are right, I will move this line to where this VC is actually presented.

File: Source/SwiftyDropbox/Platform/SwiftyDropbox_iOS/LoadingViewController.swift [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+
+import Foundation
+import UIKit
+
+/// A VC with a loading spinner at its view center.
+class LoadingViewController: UIViewController {
+    private let loadingSpinner: UIActivityIndicatorView
+
+    override init(nibName nibNameOrNil: String?, bundle nibBundleOrNil: Bundle?) {
+        loadingSpinner = UIActivityIndicatorView(style: .whiteLarge)
```
[the remark this answers is not in the document]
Comment (xuyanke) [reply Source/SwiftyDropbox/Platform/SwiftyDropbox_iOS/LoadingViewController.swift:13]: will do, thanks!

File: Source/SwiftyDropbox/Platform/SwiftyDropbox_iOS/LoadingViewController.swift [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+    }
+
+    @available(*, unavailable, message: "init(coder:) has not been implemented")
+    required init?(coder: NSCoder) {
+        fatalError("init(coder:) has not been implemented")
+    }
+
+    override func viewDidLoad() {
+        view.backgroundColor = UIColor.black.withAlphaComponent(0.4)
+        loadingSpinner.removeFromSuperview()
```
[the remark this answers is not in the document]
Comment (xuyanke) [reply Source/SwiftyDropbox/Platform/SwiftyDropbox_iOS/LoadingViewController.swift:25]: I will remove this line.

File: Source/SwiftyDropbox/Platform/SwiftyDropbox_iOS/OAuthMobile.swift (hunk 1 above)
Comment (xuyanke) [reply Source/SwiftyDropbox/Platform/SwiftyDropbox_iOS/OAuthMobile.swift:50]: 3rd party app could set the delegate so they could have control over the UX of loading.
in the sample app, I passed in nil to simply use the default implementation.

File: Source/SwiftyDropbox/Platform/SwiftyDropbox_iOS/OAuthMobile.swift (hunk 2 above)
Comment (xuyanke) [reply Source/SwiftyDropbox/Platform/SwiftyDropbox_iOS/OAuthMobile.swift:363]: yes, will make a helper to simplify the logics. thanks!

File: Source/SwiftyDropbox/Shared/Handwritten/OAuth/OAuth.swift (hunk 3 above)
Comment (xuyanke) [reply Source/SwiftyDropbox/Shared/Handwritten/OAuth/OAuth.swift:36]: OAuthManager doesn't have to own SharedApplication, so I made it weak.

File: Source/SwiftyDropbox/Platform/SwiftyDropbox_iOS/OAuthMobile.swift [hunk 7]
```diff
@@ -367,29 +367,59 @@ open class MobileSharedApplication: SharedApplication {
...
-                loadingStatusDelegate.dismissLoading()
-            } else if let loadingView = controller?.presentedViewController as? LoadingViewController {
-                loadingView.dismiss(animated: false)
-            }
+            dismissLoadingInApp()
+        }
+    }
+
+    private var isWebOAuthFlow: Bool {
+        return controller?.presentedViewController as? MobileSafariViewController != nil
```
Comment (rafe-g) [inline Source/SwiftyDropbox/Platform/SwiftyDropbox_iOS/OAuthMobile.swift:386]: you could also avoid casting by using `is`. This also works.
